CoDICE l2 angular intensity validation

lacoak21 opened the pull request:
# Change Summary

## Overview
Add additional step for angular intensities for positions 0 and 13. Add validation data comparisons. There was an issue keeping the arrays from fully validating. Me and Michael realized why and it is on the side of the CODICE team's data. I will make a new pr when they send me the new data. 

## Updated Files
- imap_processing/codice/codice_l2.py
   - Account for position 0 and 13 counts. 
- imap_processing/tests/codice/test_codice_l2.py
   - Validation tests.

tmplummer committed:
2301 hi l2   early look spacecraft frame map (#2312)

* Add more informative message when naming frame has unexpected value

* Fix some issues in L1C processing found when using flight data

* Modify L1C test based on recent change

* Fix tests
add descriptive comment

* Descriptive comment

* Update imap_processing/hi/hi_l1c.py

Co-authored-by: Greg Lucas <greg.m.lucas@gmail.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

---------

Co-authored-by: Greg Lucas <greg.m.lucas@gmail.com>
Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

sdhoyt committed:
fixed ESA binning (#2322)

laspsandoval committed:
I-ALiRT - Updated frame of reference (#2318)

greglucas committed:
MNT/ENH: Decommutate Ultra packets with derived values (#2323)

The Ultra packet definition has engineering unit conversions, so
we can call the packet_file_to_datasets with `use_derived_value=True`
to get the conversion. These are then l1b products to follow along
with the mission convention. We handle that directly in the l1a
processing by calling the routine again with a different flag. The
metadata for l1a can be re-used, except that FILLVAL, MINVAL, and
MAXVAL are all skipped for l1b because those values try to get
converted to the datatype, but the datatype has changed with enumeration
and conversions from uint8 to str or uint8 to float, etc. For ease
of use, we just skip those metadata fields for now.

lacoak21 committed:
ULTRA l1b filter out events that are during a repointing (#2324)

* filter out events that were not during the current pointing

lacoak21 committed:
add todos to remove pytest workaroudns

lacoak21 committed:
add todos to remove pytest workaroudns

lacoak21 committed:
fix indicies

tech3371 commented on imap_processing/codice/codice_l2.py:
```diff
@@ -383,8 +387,41 @@ def process_lo_angular_intensity(
+    # To account for this, we replicate the counts observed in position 0 and 10 for
+    # each esa step to either spin angles 0-11 or 12-23, depending on the pixel
+    # orientation (A/B). See section 11.2.2 of the CoDICE algorithm document
+    a_inds = np.array(
+        [pos for pos, orientation in PIXEL_ORIENTATIONS.items() if orientation == "A"]
+    )
+    b_inds = np.array(
+        [pos for pos, orientation in PIXEL_ORIENTATIONS.items() if orientation == "B"]
+    )
```
In the future, we should let L2 read in SCI LUT if you need information like this. SCI LUT json has these information and would be good to read from same source.

tech3371 committed:
CoDICE L1A Lo species refactor and CDF attrs (#2313)

tmplummer committed:
Modify build_cdf_dataset to automatically drop variables that don't have attributes defined. (#2326)

tech3371 reviewed (approved):
Looks good to me

lacoak21 committed:
ULTRA l1b flag dont filter events in a pointing  (#2330)

* instead of filtering events, use quality flags

tmplummer committed:
2251 hi lo l2 cg correction   apply liouvilles theorem (#2328)

* Add interpolation to helioframe code to ena corrections module

* Add interpolation to helioframe to Hi L2 processing

* Fix tests based on PR feedback

lacoak21 committed:
use the actual sensor and not default 45 LUTs (#2335)

tech3371 committed:
SWAPI: science time in UTC and additional checks on L2 energy lookup (#2338)

tech3371 committed:
CoDICE: New l1a angular refactor (#2336)

laspsandoval committed:
I-ALiRT - prepare for additional db (#2306)

greglucas committed:
FIX: Correct Lo bootstrap calculations with energy dimension (#2339)

The Lo bootstrap calculation was using the correct broadcasting,
but the coordinates were not aligned between the bootstrap factors
and the intensity variables. This adds the actual energy values
to the bootstrap factors we used to align them and do the proper
summation over energies.

tech3371 committed:
CoDICE: L1B species refactoring after L1A (#2333)

greglucas committed:
Lo flux correction (#2343)

* ENH: Add flux corrector to Lo processing

lacoak21 committed:
add global attrs

tech3371 committed:
CoDICE: add Lo energy_table variable (#2344)

lacoak21 committed:
fix test

tech3371 committed:
CoDICE: L1A and L1B validation of species and angular (#2347)

lacoak21 committed:
use v007
